src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointSelector.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class DynamicControllerEndpointSelector : IDisposable
+    {
+        private readonly ControllerActionEndpointDataSource _dataSource;
+        private readonly DataSourceDependentCache<ActionSelectionTable<RouteEndpoint>> _cache;
+
+        public DynamicControllerEndpointSelector(ControllerActionEndpointDataSource dataSource)
+        {
+            if (dataSource == null)
+            {
+                throw new ArgumentNullException(nameof(dataSource));
+            }
+
+            _dataSource = dataSource;
+            _cache = new DataSourceDependentCache<ActionSelectionTable<RouteEndpoint>>(dataSource, Initialize);
+        }
+
+        private ActionSelectionTable<RouteEndpoint> CurrentState => _cache.EnsureInitialized();
+
+        // This is async because the page version will be async. We don't want to put ourselves in a
+        // position where these types become public and then users start using them in sync locations.
+        public Task<IReadOnlyList<RouteEndpoint>> SelectEndpointsAsync(RouteValueDictionary values)
+        {
+            if (values == null)
+            {
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            var state = CurrentState;
```

Comment:
nit: table is much better than state IMHO.

Reply:
Fowler hates this so much, but he's wrong.

Reply:
meh.